### PR TITLE
Add complex dtype support to cast

### LINKED ISSUE #3346: Casting from real to complex lacks gradient
The operation of casting from `float32` to `complex64` lacks a gradient:
```
>>> x = tf.Variable(tf.zeros([10]))
>>> tf.gradients(tf.cast(x, dtype=tf.complex64), x)
[None]
```
Surely this is not intentional?

(FWIW, I'm on Tensorflow 0.9.0, running on the CPU on OS X.)

### PR DESCRIPTION
**Pull request by ibab**
This PR enables the `tf.cast` op to convert from/to complex numbers.
This is something that users might expect to work, as it's possible in numpy.
I've also extended the tests and gradient code.

I didn't quite know what to put as the cost for the cast operation, so I picked `NumTraits<To>::AddCost`, which is similar to what's used by `bfloat16`.

This fixes #3346.

### CODE REVIEW

**Comment by danmane** [comment]
@benoitsteiner Can you review?

**Comment by benoitsteiner** [comment]
Jenkins, test this please.

File: tensorflow/core/kernels/cast_op.h [hunk 1]
```diff
@@ -43,6 +43,111 @@ struct CastFunctor {
...
+using tensorflow::uint16;
+using tensorflow::int16;
+using tensorflow::int32;
+using tensorflow::int64;
+
+// These are seperate definitions of what should happen when we cast from/to
+// complex numbers. The actual template specializations are instantiated below.
+template<typename From, typename To>
+struct cast_from_complex_impl {
+  EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE To operator()(const From& a) const {
```
Comment (benoitsteiner) [inline tensorflow/core/kernels/cast_op.h]: Consider adding a comment stating that we're replicating the numpy behavior here.

File: tensorflow/core/kernels/cast_op.h [hunk 2]
```diff
@@ -43,6 +43,111 @@ struct CastFunctor {
...
+    return static_cast<To>(a.real());
+  }
+};
+
+template<typename From, typename To>
```
Comment (benoitsteiner) [inline tensorflow/core/kernels/cast_op.h]: The code would be simpler if you used partial specialization instead: you could for example write this to implement the casting from complex to real.
template <typename From, typename To>
struct scalar_cast_op<std::complex<From>, To> {
 EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE To operator()(const std::complex<From>& a) const {
  return static_cast<To>(a.real());
  }
};

Comment (ibab) [inline tensorflow/core/kernels/cast_op.h]: Ah, I spent a lot of time trying to make a partial specialization approach work.
But I was trying to use SFINAE to specialize for both `complex64` and `complex128` at once, which isn't really possible here.
I should have realized that I can do
```c++
template<typename From, typename To>
struct scalar_cast_op<std::complex<From>, To> {
  To operator()(const std::complex<From>& a) const {
     ...
   }
};
```
as you've suggested.
This works really well, and should make the code a lot more elegant.
I'll update the PR.

**Comment by danmane** [comment]
The test failures so far are known issues; they don't appear related to this PR.

**Comment by ibab** [comment]
I've pushed a new commit that uses partial specialization instead of macros.

**Comment by benoitsteiner** [comment]
Jenkins, test this please.

**Comment by benoitsteiner** [comment]
The test failures appear unrelated, and the CPU tests were already failing when Jenkins started to test this PR (http://ci.tensorflow.org/job/tensorflow-master-cpu/926/). However to be on the safe side I'll wait until TensorFlow tests turn green again and trigger the tests again.

**Comment by benoitsteiner** [comment]
Jenkins, test this please.

**Comment by benoitsteiner** [comment]
Jenkins, test this please.

**Pull request merged**